lib/mix/lib/mix/tasks/release.ex

```diff
@@ -1007,6 +1013,35 @@ defmodule Mix.Tasks.Release do
     release
   end
 
+  defp make_tar(release) do
+    tar_filename = "#{release.name}.tar.gz"
+    out_path = Path.join(release.path, tar_filename)
+    info(release, [:green, "* building ", :reset, out_path])
+
+    lib_dirs =
+      Enum.reduce(release.applications, [], fn {name, app_config}, acc ->
+        vsn = Keyword.fetch!(app_config, :vsn)
+        [Path.join("lib", "#{name}-#{vsn}") | acc]
+      end)
+
+    release_files =
+      Path.join(release.path, "releases")
+      |> File.ls!()
+      |> Enum.filter(&(!File.dir?(Path.join([release.path, "releases", &1]))))
+      |> Enum.map(&Path.join("releases", &1))
+
+    dirs =
+      ["bin", Path.join("releases", release.version), "erts-#{release.erts_version}"] ++
+        lib_dirs ++ release_files
+
+    files =
+      Enum.map(dirs, &{String.to_charlist(&1), String.to_charlist(Path.join(release.path, &1))})
+
+    :ok = :erl_tar.create(String.to_charlist(out_path), files, [:dereference, :compressed])
+    :ok = File.rename(out_path, Path.join(release.version_path, tar_filename))
```

Comment:
By using the path I proposed above, we don't need the rename anymore, but we will need to call `File.rm` before `:erl_tar.create`.